torch/lib/c10d/Utils.cpp

```diff
@@ -210,21 +213,27 @@ int connect(
       break;
 
     } catch (std::exception& e) {
+      // ECONNREFUSED happens if the server is not yet listening.
       if (errno == ECONNREFUSED) {
         anyRefused = true;
       }
+      // ECONNRESET happens if the server's listen backlog is exhausted.
+      if (errno == ECONNRESET) {
+        anyReset = true;
+      }
 
       // We need to move to the next address because this was not available
       // to connect or to create a socket.
       nextAddr = nextAddr->ai_next;
 
       // We have tried all addresses but could not connect to any of them.
       if (!nextAddr) {
-        if (!wait || !anyRefused) {
+        if (!wait || (!anyRefused && !anyReset)) {
           throw;
         }
         std::this_thread::sleep_for(std::chrono::seconds(1));
         anyRefused = false;
+        anyReset = false;
         nextAddr = addresses.get();
```

Comment:
does this mean we are going to do endless retries if every iteration hits `ECONNREFUSED` or `ECONNRESET`?

Reply:
Yes, that's correct. This commit doesn't change that, though.

I filed #25769 to deal with that separately.